core/types/transaction.go

```diff
@@ -488,6 +503,47 @@ func (t *TransactionsByPriceAndNonce) CurrentSize() int {
 	return len(t.heads)
 }
 
+//Forward move t to be one index behind tx, tx cant be nil
+func (t *TransactionsByPriceAndNonce) Forward(tx *Transaction) {
+	if tx == nil {
+		txTmp := t.Peek()
+		for txTmp != nil {
+			t.Shift()
+			txTmp = t.Peek()
+		}
+		return
+	}
+
+	l := len(t.heads)
+	acc, _ := Sender(t.signer, tx)
+	for i := 0; i < l; i++ {
```

Comment:
why not range `t.heads` here?

Reply:
Within this loop, it will visit `t.heads[i]` and do `t.Shift()` which will affect the length of `t.heads`, it is a dangerous signal even it is safe currently.  Can we try to move the catch-up logic outsides the loop?

Reply:
updated